aospy/automate.py

```diff
@@ -0,0 +1,247 @@
+    return [dict(zip(specs.keys(), perm)) for perm in permuter]
+
+
+def _merge_dicts(*dict_args):
+    """Merge the given dictionaries into single dict.
+
+    Given any number of dicts, shallow copy and merge into a new dict,
+    precedence goes to key value pairs in latter dicts.
+
+    From http://stackoverflow.com/a/26853961/1706640
+    """
+    result = {}
+    for dictionary in dict_args:
+        result.update(dictionary)
+    return result
+
+
+def _input_func_py2_py3():
+    """Find function for reading user input that works on Python 2 and 3.
+
+    See e.g. http://stackoverflow.com/questions/21731043/use-of-input-raw-input-in-python-2-and-3
+    """
+    try:
+        input = raw_input
+    except NameError:
+        import builtins
+        input = builtins.input
+    return input
+
+
+def _user_verify():
+    input = _input_func_py2_py3()
+    if not input("Perform these computations? [y/n] ").lower()[0] == 'y':
+        raise IOError('\n', 'Execution cancelled by user.')
+
+
+def _get_all_objs_of_type(type_, parent):
+    return set([obj for obj in parent.__dict__.values()
+                if isinstance(obj, type_)])
+
+
+class CalcSuite(object):
+    """Generate suite of Calc objects based on inputted specifications."""
+
+    _CORE_SPEC_NAMES = ('projects',
+                        'models',
+                        'runs')
+
+    _AUX_SPEC_NAMES = ('variables',
+                       'date_ranges',
+                       'regions',
+                       'input_time_intervals',
+                       'input_time_datatypes',
+                       'input_time_offsets',
+                       'input_vertical_datatypes',
+                       'output_time_intervals',
+                       'output_time_regional_reductions',
+                       'output_vertical_reductions')
+
+    _NAMES_SUITE_TO_CALC_MAPPING = {
+        'projects': 'proj',
+        'models': 'model',
+        'runs': 'run',
+        'variables': 'var',
+        'date_ranges': 'date_range',
+        'regions': 'region',
+        'input_time_intervals': 'intvl_in',
+        'input_time_datatypes': 'dtype_in_time',
+        'input_time_offsets': 'time_offset',
+        'input_vertical_datatypes': 'dtype_in_vert',
+        'output_time_intervals': 'intvl_out',
+        'output_time_regional_reductions': 'dtype_out_time',
+        'output_vertical_reductions': 'dtype_out_vert',
+    }
+
+    def __init__(self, calc_suite_specs, obj_lib):
+        """Turn all inputs into aospy-ready objects."""
+        self._specs_in = calc_suite_specs
+        self._obj_lib = obj_lib
+
+        self._projects = self._get_projects()
+
+    def _get_projects(self):
+        projs = _get_all_objs_of_type(Proj, self._obj_lib)
+        names = self._specs_in['projects']
+        return set([proj for proj in projs if proj.name in names])
+
+    def _get_regions(self):
+        return [_get_all_objs_of_type(Region, self._obj_lib)]
+
+    def _get_variables(self):
```

Comment:
Would rewriting this function to look like this be acceptable?  This would basically revert the behavior back to using the python variable name for each Var (not its name attribute).  
```python
def _get_variables(self):
    objs = getattr(self._obj_lib, 'variables', self._obj_lib).__dict__
    variables = {
        name: var for name, var in objs.items() if isinstance(var, Var)}
    names = self._specs_in['variables']
    return set([var for name, var in variables.items() if name in names])
```
This is pretty confusing though, since the name attribute is used for every other core object type (though that's also the way things are handled in the main script prior to this PR).  

To be honest I think for my particular situation / use-case, writing a similar, simpler, kind of main script to handle permutation that asked the user to specify lists of actual aospy core objects (rather than strings) would be the way to go.  It's up to you what you want to do here; I'm not suggesting we switch to that in the current PR.  I'm more considering in the medium-term writing that type of main script on my own, and potentially thinking about adding it to the aospy core later.